File: crowdshield-ai/frontend/src/types/crowd.ts
/* Types for crowd metrics, zones, and risk */

export interface ZoneMetrics {
  zone_id: string;
  bbox?: { x1: number; y1: number; x2: number; y2: number };
  person_count: number;
  density: number;
  occupancy?: number;
  velocity_avg?: number;
  flow_consistency?: number;
  growth_rate?: number;
  pressure_score?: number;
  risk_score: number;
  risk_level: RiskLevel;
}

export interface FlowArrow {
  x: number;
  y: number;
  dx: number;
  dy: number;
  magnitude: number;
  angle: number;
}

export interface CrowdSnapshot {
  type: string;
  frame_number: number;
  timestamp: number;
  total_persons: number;
  overall_cri: number;
  risk_level: RiskLevel;
  metrics: {
    density: number;
    occupancy: number;
    velocity_avg: number;
    flow_consistency: number;
    pressure_score: number;
    flow_conflict: number;
    density_growth_rate: number;
    speed_drop: number;
    counter_flow_ratio: number;
  };
  zones: ZoneMetrics[];
  detections?: Detection[];
  heatmap_url?: string;
  flow_arrows?: FlowArrow[];
  forecasts?: {
    [key: string]: {
      zones: ZoneMetrics[];
      heatmap_url?: string;
    };
  };
}

export interface Detection {
  bbox: { x1: number; y1: number; x2: number; y2: number };
  confidence: number;
  track_id?: number;
  velocity?: number;
  direction?: number;
  trajectory?: { x: number; y: number }[];
}

export interface CrowdMetric {
  id: string;
  video_id?: string;
  camera_id?: string;
  frame_number: number;
  timestamp: number;
  zone_id?: string;
  person_count: number;
  density: number;
  density_level: string;
  occupancy: number;
  velocity_avg: number;
  flow_consistency: number;
  pressure_score: number;
  risk_score: number;
  risk_level: RiskLevel;
  created_at: string;
}

export type RiskLevel = 'safe' | 'moderate' | 'high' | 'critical';

export function getRiskColor(level: RiskLevel): string {
  switch (level) {
    case 'safe': return 'var(--safe)';
    case 'moderate': return 'var(--moderate)';
    case 'high': return 'var(--high)';
    case 'critical': return 'var(--critical)';
  }
}

export function getCRILevel(cri: number): RiskLevel {
  if (cri <= 25) return 'safe';
  if (cri <= 50) return 'moderate';
  if (cri <= 75) return 'high';
  return 'critical';
}

File: crowdshield-ai/backend/app/schemas/crowd.py
"""Crowd schemas — crowd metrics, zones, risk index, and pressure."""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class CrowdMetricResponse(BaseModel):
    id: uuid.UUID
    video_id: Optional[uuid.UUID] = None
    camera_id: Optional[uuid.UUID] = None
    frame_number: int
    timestamp: float
    zone_id: Optional[str] = None
    person_count: int
    detection_confidence_avg: float
    density: float
    density_level: str
    occupancy: float
    velocity_avg: float
    flow_direction: float
    flow_consistency: float
    counter_flow_ratio: float
    density_growth_rate: float
    speed_drop: float
    pressure_score: float
    flow_conflict: float
    risk_score: float
    risk_level: str
    created_at: datetime

    model_config = {"from_attributes": True}


class ZoneMetrics(BaseModel):
    zone_id: str
    bbox: dict  # {x1, y1, x2, y2}
    person_count: int
    density: float
    occupancy: float
    velocity_avg: float
    flow_consistency: float
    growth_rate: float
    pressure_score: float
    risk_score: float
    risk_level: str


class CrowdRiskIndex(BaseModel):
    overall_cri: float  # 0-100
    risk_level: str  # safe, moderate, high, critical
    components: dict  # density, density_growth, speed_drop, pressure, flow_conflict, occupancy
    zones: list[ZoneMetrics]
    timestamp: float
    frame_number: int


class CrowdSnapshot(BaseModel):
    """Real-time crowd state pushed via WebSocket."""
    frame_number: int
    timestamp: float
    total_persons: int
    overall_cri: float
    risk_level: str
    zones: list[ZoneMetrics]
    detections: list[dict]  # [{bbox, confidence, track_id, velocity, direction}]
    heatmap_url: Optional[str] = None
    flow_arrows: Optional[list[dict]] = None
    forecasts: Optional[dict] = None


class TrackInfo(BaseModel):
    track_id: int
    bbox: dict
    velocity: float
    direction: float
    duration_frames: int
    trajectory: list[dict]  # [{x, y, frame}]
